File: paimon-python/pypaimon/write/blob_format_writer.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import struct
from collections.abc import Mapping
from typing import BinaryIO, List, Optional

try:
    from isal import isal_zlib as crc_backend
except ImportError:
    import zlib as crc_backend

from pypaimon.common.map_blob_key_serializer import create_map_blob_key_serializer
from pypaimon.schema.data_types import (
    MapType,
    is_array_blob_type,
    is_blob_file_type,
    is_map_blob_type,
)
from pypaimon.table.row.blob import Blob, BlobData, BlobDescriptor, BlobConsumer
from pypaimon.common.delta_varint_compressor import DeltaVarintCompressor


class BlobFormatWriter:
    VERSION = 1
    MAGIC_NUMBER = 1481511375
    ARRAY_VERSION = 1
    ARRAY_MAGIC_NUMBER = 1094861634
    MAP_VERSION = 1
    MAP_MAGIC_NUMBER = 0x4D424342
    NULL_LENGTH = -1
    PLACE_HOLDER_LENGTH = -2
    ARRAY_NULL_ELEMENT_LENGTH = -1
    MAP_NULL_KEY_LENGTH = -1
    MAP_NULL_VALUE_LENGTH = -1
    BUFFER_SIZE = 4096
    METADATA_SIZE = 12  # 8-byte length + 4-byte CRC

    def __init__(self, output_stream: BinaryIO,
                 blob_consumer: Optional[BlobConsumer] = None,
                 file_path: Optional[str] = None,
                 copy_buffer_size: int = BUFFER_SIZE):
        if copy_buffer_size <= 0:
            raise ValueError(
                f"BLOB copy buffer size must be positive, but was {copy_buffer_size}.")
        self.output_stream = output_stream
        self._blob_consumer = blob_consumer
        self._file_path = file_path
        self.copy_buffer_size = copy_buffer_size
        self.lengths: List[int] = []
        self.position = 0

    def add_element(self, row) -> None:
        if not hasattr(row, 'values') or len(row.values) != 1:
            raise ValueError("BlobFormatWriter only supports one field")

        blob_field_name = row.fields[0].name
        blob_field_type = row.fields[0].type
        blob_value = row.values[0]
        if isinstance(blob_field_type, MapType) and not is_map_blob_type(blob_field_type):
            raise ValueError(
                f"Map-Blob value type must be BLOB, but is {blob_field_type.value}."
            )
        if blob_value is None:
            self.lengths.append(self.NULL_LENGTH)
            if self._blob_consumer is not None:
                self._blob_consumer(blob_field_name, None)
            return

        if is_map_blob_type(blob_field_type):
            if blob_value is Blob.MAP_PLACE_HOLDER:
                self.lengths.append(self.PLACE_HOLDER_LENGTH)
                return
            self.add_blob_map(
                blob_field_name,
                blob_value,
                blob_field_type.key,
            )
            return

        if is_array_blob_type(blob_field_type):
            if blob_value is Blob.ARRAY_PLACE_HOLDER:
                self.lengths.append(self.PLACE_HOLDER_LENGTH)
                return
            self.add_blob_array(
                blob_field_name,
                blob_value,
                element_nullable=blob_field_type.element.nullable,
            )
            return

        if blob_value is Blob.PLACE_HOLDER:
            self.lengths.append(self.PLACE_HOLDER_LENGTH)
            return

        self.add_blob(blob_field_name, blob_value)

    def add_blob(self, blob_field_name: str, blob_value: Blob) -> None:
        if not isinstance(blob_value, Blob):
            raise ValueError("Field must be Blob/BlobData instance")

        previous_pos = self.position
        crc32 = 0  # Initialize CRC32

        # Write magic number
        magic_bytes = struct.pack('<I', self.MAGIC_NUMBER)  # Little endian
        crc32 = self._write_with_crc(magic_bytes, crc32)

        blob_pos, blob_length, crc32 = self._write_blob_data(blob_value, crc32)

        # Calculate total length including magic + data + metadata (length + CRC)
        bin_length = self.position - previous_pos + self.METADATA_SIZE
        self.lengths.append(bin_length)

        # Write length (8 bytes, little endian)
        length_bytes = struct.pack('<Q', bin_length)
        self.output_stream.write(length_bytes)
        self.position += 8

        # Write CRC32 (4 bytes, little endian)
        crc_bytes = struct.pack('<I', crc32 & 0xffffffff)
        self.output_stream.write(crc_bytes)
        self.position += 4

        if self._blob_consumer is not None:
            descriptor = BlobDescriptor(self._file_path, blob_pos, blob_length)
            flush = self._blob_consumer(blob_field_name, descriptor)
            if flush:
                self.output_stream.flush()

    def add_blob_array(
            self,
            blob_field_name: str,
            blob_values,
            element_nullable: bool = True) -> None:
        if isinstance(blob_values, (bytes, bytearray, str)) or not hasattr(blob_values, '__iter__'):
            raise ValueError("ARRAY<BLOB> field value must be a list or tuple of Blob values")

        blob_values = list(blob_values)
        for blob_value in blob_values:
            if blob_value is None:
                if not element_nullable:
                    raise ValueError(
                        f"ARRAY<BLOB> field '{blob_field_name}' does not allow null elements"
                    )
                continue
            if not isinstance(blob_value, Blob):
                raise ValueError("ARRAY<BLOB> elements must be Blob/BlobData instances or None")
            if blob_value is Blob.PLACE_HOLDER or blob_value is Blob.ARRAY_PLACE_HOLDER:
                raise ValueError("ARRAY<BLOB> elements do not support placeholders")

        previous_pos = self.position
        crc32 = 0

        crc32 = self._write_with_crc(struct.pack('<I', self.MAGIC_NUMBER), crc32)
        crc32 = self._write_with_crc(struct.pack('<I', self.ARRAY_MAGIC_NUMBER), crc32)
        crc32 = self._write_with_crc(struct.pack('<B', self.ARRAY_VERSION), crc32)
        crc32 = self._write_with_crc(struct.pack('<I', len(blob_values)), crc32)

        element_lengths = []
        flush = False
        for blob_value in blob_values:
            if blob_value is None:
                element_lengths.append(self.ARRAY_NULL_ELEMENT_LENGTH)
                continue

            blob_pos, blob_length, crc32 = self._write_blob_data(blob_value, crc32)
            element_lengths.append(blob_length)
            if self._blob_consumer is not None:
                descriptor = BlobDescriptor(self._file_path, blob_pos, blob_length)
                flush = self._blob_consumer(blob_field_name, descriptor) or flush

        element_index_bytes = DeltaVarintCompressor.compress(element_lengths)
        crc32 = self._write_with_crc(element_index_bytes, crc32)
        crc32 = self._write_with_crc(struct.pack('<I', len(element_index_bytes)), crc32)

        bin_length = self.position - previous_pos + self.METADATA_SIZE
        self.lengths.append(bin_length)
        self.output_stream.write(struct.pack('<Q', bin_length))
        self.position += 8
        self.output_stream.write(struct.pack('<I', crc32 & 0xffffffff))
        self.position += 4

        if flush:
            self.output_stream.flush()

    def add_blob_map(self, blob_field_name: str, blob_values, key_type) -> None:
        entries = self._map_entries(blob_values)
        if len(entries) > 0x7fffffff:
            raise ValueError(f"Invalid MAP<X, BLOB> entry count: {len(entries)}")

        key_serializer = create_map_blob_key_serializer(key_type)
        key_bytes = []
        for key, _ in entries:
            if key is None:
                key_bytes.append(None)
                continue
            serialized = key_serializer.serialize(key)
            if len(serialized) > 0x7fffffff:
                raise ValueError(f"MAP<X, BLOB> key is too large: {len(serialized)}")
            key_bytes.append(serialized)

        for _, blob_value in entries:
            if blob_value is None:
                continue
            if (
                blob_value is Blob.PLACE_HOLDER
                or blob_value is Blob.ARRAY_PLACE_HOLDER
                or blob_value is Blob.MAP_PLACE_HOLDER
            ):
                raise ValueError("MAP<X, BLOB> values do not support placeholders")
            if not isinstance(blob_value, Blob):
                raise ValueError("MAP<X, BLOB> values must be Blob/BlobData instances or None")

        previous_pos = self.position
        crc32 = 0
        crc32 = self._write_with_crc(struct.pack('<I', self.MAGIC_NUMBER), crc32)
        crc32 = self._write_with_crc(struct.pack('<I', self.MAP_MAGIC_NUMBER), crc32)
        crc32 = self._write_with_crc(struct.pack('<B', self.MAP_VERSION), crc32)
        crc32 = self._write_with_crc(struct.pack('<I', len(entries)), crc32)

        key_lengths = []
        for serialized in key_bytes:
            if serialized is None:
                key_lengths.append(self.MAP_NULL_KEY_LENGTH)
            else:
                key_lengths.append(len(serialized))
                crc32 = self._write_with_crc(serialized, crc32)

        value_lengths = []
        flush = False
        for _, blob_value in entries:
            if blob_value is None:
                value_lengths.append(self.MAP_NULL_VALUE_LENGTH)
                continue
            blob_pos, blob_length, crc32 = self._write_blob_data(blob_value, crc32)
            value_lengths.append(blob_length)
            if self._blob_consumer is not None:
                descriptor = BlobDescriptor(self._file_path, blob_pos, blob_length)
                flush = self._blob_consumer(blob_field_name, descriptor) or flush

        key_index_bytes = DeltaVarintCompressor.compress(key_lengths)
        value_index_bytes = DeltaVarintCompressor.compress(value_lengths)
        crc32 = self._write_with_crc(key_index_bytes, crc32)
        crc32 = self._write_with_crc(value_index_bytes, crc32)
        crc32 = self._write_with_crc(struct.pack('<I', len(key_index_bytes)), crc32)
        crc32 = self._write_with_crc(struct.pack('<I', len(value_index_bytes)), crc32)

        bin_length = self.position - previous_pos + self.METADATA_SIZE
        self.lengths.append(bin_length)
        crc32 = self._write_with_crc(struct.pack('<Q', bin_length), crc32)
        self.output_stream.write(struct.pack('<I', crc32 & 0xffffffff))
        self.position += 4

        if flush:
            self.output_stream.flush()

    def _write_blob_data(self, blob_value: Blob, crc32: int):
        blob_pos = self.position

        if isinstance(blob_value, BlobData):
            data = blob_value.to_data()
            crc32 = self._write_with_crc(data, crc32)
        else:
            stream = blob_value.new_input_stream()
            try:
                chunk = stream.read(self.copy_buffer_size)
                while chunk:
                    crc32 = self._write_with_crc(chunk, crc32)
                    chunk = stream.read(self.copy_buffer_size)
            finally:
                stream.close()

        return blob_pos, self.position - blob_pos, crc32

    def _write_with_crc(self, data: bytes, crc32: int) -> int:
        crc32 = crc_backend.crc32(data, crc32)
        self.output_stream.write(data)
        self.position += len(data)
        return crc32

    def write_value(self, col_data, fields, uri_reader_factory=None) -> None:
        from pypaimon.table.row.generic_row import GenericRow
        from pypaimon.table.row.row_kind import RowKind

        is_map_blob = is_map_blob_type(fields[0].type)
        is_blob = is_blob_file_type(fields[0].type) or is_map_blob

        if col_data is None:
            if not is_blob:
                raise RuntimeError("Null values are only supported for BLOB type fields")
            self.lengths.append(self.NULL_LENGTH)
            return

        if is_map_blob:
            row_values = [self._to_blob_map(col_data, uri_reader_factory)]
        elif is_array_blob_type(fields[0].type):
            row_values = [self._to_blob_array(col_data, uri_reader_factory)]
        elif is_blob:
            row_values = [self._to_blob(col_data, uri_reader_factory)]
        else:
            row_values = [col_data]

        row = GenericRow(row_values, fields, RowKind.INSERT)
        self.add_element(row)

    @staticmethod
    def _to_blob(col_data, uri_reader_factory=None) -> Blob:
        if col_data is Blob.PLACE_HOLDER:
            return Blob.PLACE_HOLDER
        if hasattr(col_data, 'as_py'):
            col_data = col_data.as_py()
        if isinstance(col_data, str):
            col_data = col_data.encode('utf-8')
        if isinstance(col_data, bytearray):
            col_data = bytes(col_data)
        if isinstance(col_data, Blob):
            return col_data
        if isinstance(col_data, bytes):
            if BlobDescriptor.is_blob_descriptor(col_data):
                if uri_reader_factory is None:
                    raise RuntimeError("uri_reader_factory is required for BlobDescriptor bytes.")
                descriptor = BlobDescriptor.deserialize(col_data)
                uri_reader = uri_reader_factory.create(descriptor.uri)
                return Blob.from_descriptor(uri_reader, descriptor)
            return BlobData(col_data)
        raise RuntimeError(
            "Blob field value must be bytes/blob or serialized BlobDescriptor bytes."
        )

    @classmethod
    def _to_blob_array(cls, col_data, uri_reader_factory=None):
        if col_data is Blob.ARRAY_PLACE_HOLDER:
            return Blob.ARRAY_PLACE_HOLDER
        if hasattr(col_data, 'as_py'):
            col_data = col_data.as_py()
        if col_data is None:
            return None
        if isinstance(col_data, (bytes, bytearray, str)) or not hasattr(col_data, '__iter__'):
            raise RuntimeError("ARRAY<BLOB> field value must be a list or tuple.")
        result = []
        for element in col_data:
            if element is None:
                result.append(None)
                continue
            if element is Blob.PLACE_HOLDER or element is Blob.ARRAY_PLACE_HOLDER:
                raise RuntimeError("ARRAY<BLOB> elements do not support placeholders.")
            result.append(cls._to_blob(element, uri_reader_factory))
        return result

    @classmethod
    def _to_blob_map(cls, col_data, uri_reader_factory=None):
        if col_data is Blob.MAP_PLACE_HOLDER:
            return Blob.MAP_PLACE_HOLDER
        if hasattr(col_data, 'as_py'):
            col_data = col_data.as_py()
        if col_data is None:
            return None

        result = []
        for key, value in cls._map_entries(col_data):
            if value is None:
                result.append((key, None))
                continue
            if (
                value is Blob.PLACE_HOLDER
                or value is Blob.ARRAY_PLACE_HOLDER
                or value is Blob.MAP_PLACE_HOLDER
            ):
                raise RuntimeError("MAP<X, BLOB> values do not support placeholders.")
            result.append((key, cls._to_blob(value, uri_reader_factory)))
        return result

    @staticmethod
    def _map_entries(blob_values):
        if isinstance(blob_values, Mapping):
            return list(blob_values.items())
        if isinstance(blob_values, (bytes, bytearray, str)) or not hasattr(
            blob_values, '__iter__'
        ):
            raise ValueError("MAP<X, BLOB> field value must be a mapping or key/value pairs")

        entries = []
        for entry in blob_values:
            if not isinstance(entry, (list, tuple)) or len(entry) != 2:
                raise ValueError("MAP<X, BLOB> entries must be key/value pairs")
            entries.append((entry[0], entry[1]))
        return entries

    def reach_target_size(self, target_size: int) -> bool:
        return self.position >= target_size

    def close(self) -> None:
        index_bytes = DeltaVarintCompressor.compress(self.lengths)
        self.output_stream.write(index_bytes)

        # Write header (index length + version)
        header = struct.pack('<I', len(index_bytes))  # Index length (4 bytes, little endian)
        header += struct.pack('<B', self.VERSION)  # Version (1 byte)
        self.output_stream.write(header)

        # Flush and close
        if hasattr(self.output_stream, 'flush'):
            self.output_stream.flush()
        if hasattr(self.output_stream, 'close'):
            self.output_stream.close()

File: paimon-format/src/main/java/org/apache/paimon/format/blob/BlobFormatWriter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.format.blob;

import org.apache.paimon.data.Blob;
import org.apache.paimon.data.BlobArrayPlaceholder;
import org.apache.paimon.data.BlobConsumer;
import org.apache.paimon.data.BlobDescriptor;
import org.apache.paimon.data.BlobFetchMetricReporter;
import org.apache.paimon.data.BlobPlaceholder;
import org.apache.paimon.data.BlobRef;
import org.apache.paimon.data.BlobReuseSource;
import org.apache.paimon.data.InternalArray;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.format.FileAwareFormatWriter;
import org.apache.paimon.format.FormatWriter;
import org.apache.paimon.fs.Path;
import org.apache.paimon.fs.PositionOutputStream;
import org.apache.paimon.fs.SeekableInputStream;
import org.apache.paimon.rest.HttpClientUtils;
import org.apache.paimon.types.DataTypeRoot;
import org.apache.paimon.types.RowType;
import org.apache.paimon.utils.DeltaVarintCompressor;
import org.apache.paimon.utils.LongArrayList;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;

import java.io.EOFException;
import java.io.IOException;
import java.util.zip.CRC32;

import static org.apache.paimon.utils.Preconditions.checkArgument;
import static org.apache.paimon.utils.StreamUtils.intToLittleEndian;
import static org.apache.paimon.utils.StreamUtils.longToLittleEndian;

/** {@link FormatWriter} for blob file. */
public class BlobFormatWriter implements FileAwareFormatWriter {

    private static final Logger LOG = LoggerFactory.getLogger(BlobFormatWriter.class);

    public static final byte VERSION = 1;
    public static final int MAGIC_NUMBER = 1481511375;
    public static final byte[] MAGIC_NUMBER_BYTES = intToLittleEndian(MAGIC_NUMBER);
    public static final byte ARRAY_VERSION = 1;
    public static final int ARRAY_MAGIC_NUMBER = 1094861634;
    public static final byte[] ARRAY_MAGIC_NUMBER_BYTES = intToLittleEndian(ARRAY_MAGIC_NUMBER);
    public static final long NULL_LENGTH = -1L;
    public static final long PLACE_HOLDER_LENGTH = -2L;
    public static final long ARRAY_NULL_ELEMENT_LENGTH = -1L;
    public static final int DEFAULT_COPY_BUFFER_SIZE = 4 * 1024;

    private final PositionOutputStream out;
    @Nullable private final BlobConsumer writeConsumer;
    private final BlobFetchMetricReporter blobFetchMetricReporter;
    private final String blobFieldName;
    private final boolean writeNullOnMissingFile;
    private final boolean writeNullOnFetchFailure;
    private final BlobElementWriter elementWriter;
    private final CRC32 crc32;
    private final byte[] tmpBuffer;
    private final LongArrayList lengths;

    private String pathString;

    // Reuses one source stream across consecutive references into the same file, so compaction of
    // descriptor/remote blobs opens each source once instead of once per blob.
    private final BlobReuseSource reuseSource = new BlobReuseSource();

    public BlobFormatWriter(
            PositionOutputStream out,
            @Nullable BlobConsumer writeConsumer,
            RowType type,
            boolean writeNullOnMissingFile,
            boolean writeNullOnFetchFailure,
            BlobFetchMetricReporter blobFetchMetricReporter,
            int copyBufferSize) {
        checkArgument(
                copyBufferSize > 0,
                "BLOB copy buffer size must be positive, but was %s.",
                copyBufferSize);
        this.out = out;
        this.writeConsumer = writeConsumer;
        this.blobFetchMetricReporter = blobFetchMetricReporter;
        this.writeNullOnMissingFile = writeNullOnMissingFile;
        this.writeNullOnFetchFailure = writeNullOnFetchFailure;
        checkArgument(type.getFieldCount() == 1, "BlobFormatWriter only support one field.");
        this.blobFieldName = type.getFieldNames().get(0);
        this.elementWriter =
                type.getTypeAt(0).getTypeRoot() == DataTypeRoot.ARRAY
                        ? new ArrayBlobElementWriter()
                        : new RawBlobElementWriter();
        this.crc32 = new CRC32();
        this.tmpBuffer = new byte[copyBufferSize];
        this.lengths = new LongArrayList(16);
    }

    @Override
    public void setFile(Path file) {
        this.pathString = file.toString();
    }

    @Override
    public boolean deleteFileUponAbort() {
        return writeConsumer == null;
    }

    @Override
    public void addElement(InternalRow element) throws IOException {
        checkArgument(element.getFieldCount() == 1, "BlobFormatWriter only support one field.");
        elementWriter.addElement(element);
    }

    private interface BlobElementWriter {
        void addElement(InternalRow element) throws IOException;
    }

    private class RawBlobElementWriter implements BlobElementWriter {

        @Override
        public void addElement(InternalRow element) throws IOException {
            if (element.isNullAt(0)) {
                recordPreCheckedMissingFileNull(element);
                writeNullElement();
                return;
            }

            Blob blob;
            try {
                blob = element.getBlob(0);
            } catch (RuntimeException e) {
                if (shouldWriteNullOnFetchFailure(e)) {
                    logWriteNullOnFetchFailure(e, null);
                    blobFetchMetricReporter.recordFetchFailureNullWritten(e);
                    writeNullElement();
                    return;
                }
                blobFetchMetricReporter.recordFetchFailure(e);
                throw e;
            }
            if (blob == BlobPlaceholder.INSTANCE) {
                lengths.add(PLACE_HOLDER_LENGTH);
                return;
            }

            addBlob(blob);
        }
    }

    private class ArrayBlobElementWriter implements BlobElementWriter {

        @Override
        public void addElement(InternalRow element) throws IOException {
            if (element.isNullAt(0)) {
                writeNullElement();
                return;
            }

            InternalArray array = element.getArray(0);
            if (array == BlobArrayPlaceholder.INSTANCE) {
                lengths.add(PLACE_HOLDER_LENGTH);
                return;
            }

            addBlobArray(array);
        }
    }

    private void addBlob(Blob blob) throws IOException {
        BlobCopySource source = prepareBlobSource(blob);
        if (source == null) {
            writeNullElement();
            return;
        }
        long previousPos = out.getPos();
        crc32.reset();
        write(MAGIC_NUMBER_BYTES);

        BlobDescriptor descriptor = writeBlobData(source);

        long binLength = out.getPos() - previousPos + 12;
        lengths.add(binLength);
        byte[] lenBytes = longToLittleEndian(binLength);
        write(lenBytes);
        int crcValue = (int) crc32.getValue();
        out.write(intToLittleEndian(crcValue));

        if (writeConsumer != null) {
            boolean flush = writeConsumer.accept(blobFieldName, descriptor);
            if (flush) {
                out.flush();
            }
        }
        blobFetchMetricReporter.recordSuccess(descriptor.length());
    }

    private void addBlobArray(InternalArray array) throws IOException {
        long previousPos = out.getPos();
        crc32.reset();

        write(MAGIC_NUMBER_BYTES);

        write(ARRAY_MAGIC_NUMBER_BYTES);
        write(new byte[] {ARRAY_VERSION});
        write(intToLittleEndian(array.size()));

        long[] elementLengths = new long[array.size()];
        boolean flush = false;
        for (int i = 0; i < array.size(); i++) {
            if (array.isNullAt(i)) {
                elementLengths[i] = ARRAY_NULL_ELEMENT_LENGTH;
                continue;
            }

            Blob blob = getArrayBlob(array, i);
            if (blob == null) {
                elementLengths[i] = ARRAY_NULL_ELEMENT_LENGTH;
                continue;
            }
            BlobCopySource source = prepareBlobSource(blob);
            if (source == null) {
                elementLengths[i] = ARRAY_NULL_ELEMENT_LENGTH;
                continue;
            }
            BlobDescriptor descriptor = writeBlobData(source);
            elementLengths[i] = descriptor.length();
            if (writeConsumer != null) {
                flush |= writeConsumer.accept(blobFieldName, descriptor);
            }
            blobFetchMetricReporter.recordSuccess(descriptor.length());
        }

        byte[] elementIndexBytes = DeltaVarintCompressor.compress(elementLengths);
        write(elementIndexBytes);
        write(intToLittleEndian(elementIndexBytes.length));

        long binLength = out.getPos() - previousPos + 12;
        lengths.add(binLength);
        write(longToLittleEndian(binLength));
        int crcValue = (int) crc32.getValue();
        out.write(intToLittleEndian(crcValue));

        if (flush) {
            out.flush();
        }
    }

    @Nullable
    private Blob getArrayBlob(InternalArray array, int pos) {
        try {
            return array.getBlob(pos);
        } catch (RuntimeException e) {
            if (shouldWriteNullOnFetchFailure(e)) {
                logWriteNullOnFetchFailure(e, null);
                blobFetchMetricReporter.recordFetchFailureNullWritten(e);
                return null;
            }
            blobFetchMetricReporter.recordFetchFailure(e);
            throw e;
        }
    }

    /**
     * Prepares the byte source for a blob, or returns {@code null} to write it as NULL (missing
     * file / fetch failure). An exact {@link BlobRef} with known length reuses one source stream
     * (bounded to the descriptor); other blobs open their own stream and read until EOF.
     */
    @Nullable
    private BlobCopySource prepareBlobSource(Blob blob) throws IOException {
        // Exact class only: subclasses may override newInputStream() and must not be bypassed.
        if (blob != null && blob.getClass() == BlobRef.class) {
            BlobRef ref = (BlobRef) blob;
            long length = ref.toDescriptor().length();
            if (length >= 0) {
                // Position/release the previous source first; its cleanup error is not this blob's
                // fetch failure and must not be turned into a NULL write.
                reuseSource.prepareFor(ref);
                SeekableInputStream source = openStream(ref, () -> reuseSource.openBounded(ref));
                if (source == null) {
                    return null;
                }
                return new BlobCopySource(source, length);
            }
        }
        SeekableInputStream in = openStream(blob, blob::newInputStream);
        if (in == null) {
            return null;
        }
        return new BlobCopySource(in, -1L);
    }

    @FunctionalInterface
    private interface StreamOpener {
        SeekableInputStream open() throws IOException;
    }

    /**
     * Opens a source stream via {@code opener}, converting missing-file / fetch-failure errors into
     * a {@code null} result when the writer is configured to write NULL for such blobs.
     */
    @Nullable
    private SeekableInputStream openStream(Blob blob, StreamOpener opener) throws IOException {
        try {
            return opener.open();
        } catch (IOException | RuntimeException e) {
            if (writeNullOnError(blob, e)) {
                return null;
            }
            throw e;
        }
    }

    /** Applies the write-null-on-missing/fetch-failure policy; returns true to write NULL. */
    private boolean writeNullOnError(Blob blob, Throwable e) {
        if (writeNullOnMissingFile && HttpClientUtils.isNotFoundError(e)) {
            LOG.warn(
                    "Failed to open blob from {} (HTTP 404), writing NULL for BLOB field {}.",
                    blobUri(blob),
                    blobFieldName,
                    e);
            blobFetchMetricReporter.recordMissingFileNullWritten(true);
            return true;
        }
        if (shouldWriteNullOnFetchFailure(e)) {
            logWriteNullOnFetchFailure(e, blob);
            blobFetchMetricReporter.recordFetchFailureNullWritten(e);
            return true;
        }
        blobFetchMetricReporter.recordFetchFailure(e);
        return false;
    }

    private BlobDescriptor writeBlobData(BlobCopySource source) throws IOException {
        long blobPos = out.getPos();
        if (source.reused()) {
            // Bounded stream from reuseSource; copy exactly descriptor.length bytes.
            try {
                copyExactly(source.stream(), source.length());
            } catch (IOException | RuntimeException e) {
                blobFetchMetricReporter.recordFetchFailure(e);
                // Source is at an unknown position now; drop it so the next blob reopens.
                reuseSource.discardQuietly();
                throw e;
            }
        } else {
            try (SeekableInputStream stream = source.stream()) {
                int bytesRead = stream.read(tmpBuffer);
                while (bytesRead >= 0) {
                    write(tmpBuffer, bytesRead);
                    bytesRead = stream.read(tmpBuffer);
                }
            } catch (IOException | RuntimeException e) {
                blobFetchMetricReporter.recordFetchFailure(e);
                throw e;
            }
        }

        return new BlobDescriptor(pathString, blobPos, out.getPos() - blobPos);
    }

    /** Copies exactly {@code length} bytes from {@code stream}, throwing on premature EOF. */
    private void copyExactly(SeekableInputStream stream, long length) throws IOException {
        long remaining = length;
        while (remaining > 0) {
            int toRead = (int) Math.min(tmpBuffer.length, remaining);
            int bytesRead = stream.read(tmpBuffer, 0, toRead);
            if (bytesRead < 0) {
                throw new EOFException(
                        String.format(
                                "Unexpected EOF while copying BLOB payload for field %s: expected %d "
                                        + "bytes but source ended %d bytes early.",
                                blobFieldName, length, remaining));
            }
            if (bytesRead == 0) {
                throw new IOException(
                        "Source returned 0 bytes while copying BLOB payload for field "
                                + blobFieldName);
            }
            write(tmpBuffer, bytesRead);
            remaining -= bytesRead;
        }
    }

    /** The byte source of a single blob payload to be copied into the blob file. */
    private static final class BlobCopySource {

        private final SeekableInputStream stream;
        private final long length;

        private BlobCopySource(SeekableInputStream stream, long length) {
            this.stream = stream;
            this.length = length;
        }

        /**
         * Whether the payload is copied from a reused, pre-positioned source (exact-length read).
         */
        private boolean reused() {
            return length >= 0;
        }

        private SeekableInputStream stream() {
            return stream;
        }

        private long length() {
            return length;
        }
    }

    private boolean shouldWriteNullOnFetchFailure(Throwable e) {
        return writeNullOnFetchFailure && !HttpClientUtils.isNotFoundError(e);
    }

    private void logWriteNullOnFetchFailure(Throwable e, @Nullable Blob blob) {
        Integer statusCode = HttpClientUtils.getHttpStatusCode(e);
        if (statusCode != null) {
            LOG.warn(
                    "Failed to open blob from {} (HTTP {}), writing NULL for BLOB field {}.",
                    blobUri(blob),
                    statusCode,
                    blobFieldName,
                    e);
        } else if (HttpClientUtils.isInvalidUriException(e)) {
            LOG.warn(
                    "Invalid blob URI {} while opening blob, writing NULL for BLOB field {}.",
                    blobUri(blob),
                    blobFieldName,
                    e);
        } else {
            LOG.warn(
                    "Failed to open blob from {} due to fetch failure, writing NULL for BLOB field {}.",
                    blobUri(blob),
                    blobFieldName,
                    e);
        }
    }

    private void recordPreCheckedMissingFileNull(InternalRow element) {
        if (!writeNullOnMissingFile) {
            return;
        }
        Blob blob = element.getBlob(0);
        if (blob instanceof BlobRef) {
            BlobDescriptor descriptor = ((BlobRef) blob).toDescriptor();
            blobFetchMetricReporter.recordMissingFileNullWritten(isHttpUri(descriptor.uri()));
        }
    }

    private void writeNullElement() throws IOException {
        lengths.add(NULL_LENGTH);
        if (writeConsumer != null) {
            writeConsumer.accept(blobFieldName, null);
        }
    }

    private static String blobUri(@Nullable Blob blob) {
        if (blob instanceof BlobRef) {
            return ((BlobRef) blob).toDescriptor().uri();
        }
        return "unknown";
    }

    private static boolean isHttpUri(String uri) {
        return uri.regionMatches(true, 0, "http://", 0, "http://".length())
                || uri.regionMatches(true, 0, "https://", 0, "https://".length());
    }

    private void write(byte[] bytes) throws IOException {
        write(bytes, bytes.length);
    }

    private void write(byte[] bytes, int length) throws IOException {
        crc32.update(bytes, 0, length);
        out.write(bytes, 0, length);
    }

    @Override
    public boolean reachTargetSize(boolean suggestedCheck, long targetSize) throws IOException {
        // check target size every record
        // Each blob is very large, so the cost of check is not high
        return out.getPos() >= targetSize;
    }

    @Override
    public void close() throws IOException {
        Throwable primary = null;
        try {
            // index
            byte[] indexBytes = DeltaVarintCompressor.compress(lengths.toArray());
            out.write(indexBytes);
            // header
            out.write(intToLittleEndian(indexBytes.length));
            out.write(VERSION);
        } catch (RuntimeException | Error | IOException e) {
            primary = e;
            throw e;
        } finally {
            // Surface the footer error as primary and attach a source-close error as suppressed.
            if (primary == null) {
                reuseSource.close();
            } else {
                try {
                    reuseSource.close();
                } catch (RuntimeException | Error | IOException suppressed) {
                    primary.addSuppressed(suppressed);
                }
            }
        }
    }
}
